### ISSUE TITLE
Evaluating scores ONLY at a particular position (?) restricted softmax

### ISSUE THREAD

**Issue by miguelballesteros** [clab/dynet#402]
Hi all,

 I want to run a restricted softmax in which the set of possible candidates is huge ~200k, but when I run it, I normally just care about ~10 candidates.

Running incremental_forward with this cand_size is very slow,  since even if the softmax is restricted, incremental_forward will attempt to iterate over the entire thing.

The code of what I just described would be:
```
1. Expression r_t2 = affine_transform({candbias, p2c, nlp_t2});
2. Expression adistecand = log_softmax(r_t2, candsInt);   //candsInt is of size 10 (for example)
3. **vector<float> adistcand = as_vector(hg->incremental_forward(adistecand));**
4. Find max in adistcand.
```
(candbias is of size 200k, p2c is of size 200kx100 and nlp_t2 is the summary of an LSTM after a relu or similar.)

how can I generate the matrix to just select the rows I need and building the matrix just to contain those rows? any example I can look up? any ideas about how to do the restricted softmax in a fast way?

thanks a lot,
Miguel

**Comment by miguelballesteros** [comment]
Just an update, 
I tried picking the candidates from the vector first, but it is the same slow behavior when I run incremental_forward.

vector<Expression> v(candsInt.size());
for (unsigned i = 0; i < v.size(); ++i)
  v[i] = pick(r_t2, candsInt[i]);
Expression logits = concatenate(v);
Expression adistecand = log_softmax(logits);
vector<float> adistcand = as_vector(hg->incremental_forward(adistecand));

and I also tried to generate the matrix myself, but yet again the same slow behavior when I run incremental_forward.

Expression W=select_rows(p2c,candsInt);
vector<Expression> v(candsInt.size());
for (unsigned i = 0; i < v.size(); ++i)
        v[i] = pick(candbias, candsInt[i]);
Expression BIAS=concatenate(v);
Expression x = affine_transform({BIAS, W, nlp_t2});
Expression adistecand = log_softmax(x);     
vector<float> adistcand = as_vector(hg->incremental_forward(adistecand));

any ideas why this happens? am I doing something wrong?

It would be ideal to be able to **double score=pick_score(adistecand, candsInt[i]);** not sure if this is possible.

Thanks in advance,
Miguel

**Comment by miguelballesteros** [comment]
I was doing something stupid some lines above that was making my graph much bigger than I expected so no issue at all (I'm closing this). 
The two solutions in the last comment are nice approaches to solve this issue (you have a cand size huge, and you only care about a few in each timestep)

M.
